pingpong/lti/allowlist.py

```
@@ -148,6 +149,36 @@ def allow_http_for_lti_host(host: str, context: LTIHostValidationContext) -> boo
+        raise LTIUrlValidationError(f"Invalid URL path for {field_name}")
+    return quote(path, safe="/:=+!*'(),@&~")
+
+
+def _sanitize_url_query(query: str, field_name: str) -> str:
+    """Re-encode a URL query string to produce a fresh, untainted string.
+
+    Rejects query strings containing control characters.  The returned string
+    is reconstructed via :func:`urlencode` so it is no longer tainted.
+    """
+    if _UNSAFE_PATH_RE.search(query):
+        raise LTIUrlValidationError(f"Invalid URL query for {field_name}")
```

Ruff (0.15.2)

[warning] Avoid specifying long messages outside the exception class

(TRY003)